catboost/catboost_integration.py

```diff
@@ -0,0 +1,79 @@
+"""
+Optuna example that demonstrates a pruner for CatBoost.
+
+In this example, we optimize the validation accuracy of cancer detection using CatBoost.
+We optimize both the choice of booster model and their hyperparameters. Throughout
+training of models, a pruner observes intermediate results and stop unpromising trials.
+
+You can run this example as follows:
+    $ python catboost_integration.py
+
+"""
+
+import numpy as np
+import optuna
+
+import catboost as cb
+from sklearn.datasets import load_breast_cancer
+from sklearn.metrics import accuracy_score
+from sklearn.model_selection import train_test_split
+
+
+def objective(trial):
+    data, target = load_breast_cancer(return_X_y=True)
+    train_x, valid_x, train_y, valid_y = train_test_split(data, target, test_size=0.3)
+
+    param = {
+        "objective": trial.suggest_categorical("objective", ["Logloss", "CrossEntropy"]),
+        "colsample_bylevel": trial.suggest_float("colsample_bylevel", 0.01, 0.1),
+        "depth": trial.suggest_int("depth", 1, 12),
+        "boosting_type": trial.suggest_categorical("boosting_type", ["Ordered", "Plain"]),
+        "bootstrap_type": trial.suggest_categorical(
+            "bootstrap_type", ["Bayesian", "Bernoulli", "MVS"]
+        ),
+        "used_ram_limit": "3gb",
+        "eval_metric": "AUC",
+    }
+
+    if param["bootstrap_type"] == "Bayesian":
+        param["bagging_temperature"] = trial.suggest_float("bagging_temperature", 0, 10)
+    elif param["bootstrap_type"] == "Bernoulli":
+        param["subsample"] = trial.suggest_float("subsample", 0.1, 1)
+
+    gbm = cb.CatBoostClassifier(**param)
+
+    pruning_callback = optuna.integration.CatBoostPruningCallback(trial, "AUC", "validation")
+
+    gbm.fit(
+        train_x,
+        train_y,
+        eval_set=[(valid_x, valid_y)],
+        verbose=0,
+        early_stopping_rounds=100,
+        callbacks=[pruning_callback],
+    )
+    pruning_callback.check_pruned()
+
+    preds = gbm.predict(valid_x)
+    pred_labels = np.rint(preds)
+    accuracy = accuracy_score(valid_y, pred_labels)
+
+    return accuracy
```

Comment:
Maybe silly comment: In my understanding, AUC score on the validation dataset should be returned for consistency to the training objective. Otherwise, how about optimizing the accuracy of the training dataset. What do you think?